File: Backend/Mono/Nonagon.Modular.DynamicForm/Template/FormElement.cs
using System;

namespace Nonagon.Modular.DynamicForm.Template
{
	/// <summary>
	/// Form element base class.
	/// </summary>
	public abstract class FormElement : IFormElement
	{
		/// <summary>
		/// Gets or sets the width of the element.
		/// </summary>
		/// <value>The width.</value>
		public FormLayoutWidth Width { get; set; }
		
		/// <summary>
		/// Gets or sets the height of the element
		/// </summary>
		/// <value>The height.</value>
		public FormLayoutHeight Height { get; set; }
		
		/// <summary>
		/// Gets or sets the margin of the element.
		/// </summary>
		/// <value>The margin.</value>
		public FormLayoutThickness Margin { get; set; }
		
		/// <summary>
		/// Gets or sets the padding of the element.
		/// </summary>
		/// <value>The padding.</value>
		public FormLayoutThickness Padding { get; set; }
	}
}

File: Frontend/GWT/gwt-modular-dynamicform/src/com/nonagon/modular/dynamicform/client/template/FormElement.java
package com.nonagon.modular.dynamicform.client.template;

import com.google.gwt.core.client.JavaScriptObject;
import com.nonagon.modular.dynamicform.client.FormLayoutHeight;
import com.nonagon.modular.dynamicform.client.FormLayoutThickness;
import com.nonagon.modular.dynamicform.client.FormLayoutWidth;

public abstract class FormElement extends JavaScriptObject {
	
	protected FormElement() {}
	
	public final native FormLayoutThickness getMargin()/*-{ return this.Margin; }-*/;
	public final native void setMargin(FormLayoutThickness value)/*-{ this.Margin = value; }-*/;

	public final native FormLayoutThickness getPadding()/*-{ return this.Padding; }-*/;
	public final native void getPadding(FormLayoutThickness value)/*-{ this.Padding = value; }-*/;

	public final FormLayoutWidth getWidth() {
		
		String fieldLayoutWidth = jsniGetWidth();
		return FormLayoutWidth.fromString(fieldLayoutWidth);
	}
	
	public final void setWidth(FormLayoutWidth width) {
		
		String fieldLayoutWidth = FormLayoutWidth.fromEnum(width);
		jsniSetWidth(fieldLayoutWidth);
	}
	
	private final native String jsniGetWidth()/*-{ return this.Width; }-*/;
	private final native void jsniSetWidth(String value)/*-{ this.Width = value; }-*/;
	
	public final FormLayoutHeight getHeight() {
		
		String fieldLayoutHeight = jsniGetHeight();
		return FormLayoutHeight.fromString(fieldLayoutHeight);
	}
	
	public final void setHeight(FormLayoutHeight height) {
		
		String fieldLayoutHeight = FormLayoutHeight.fromEnum(height);
		jsniSetHeight(fieldLayoutHeight);
	}
	
	private final native String jsniGetHeight()/*-{ return this.Height; }-*/;
	private final native void jsniSetHeight(String value)/*-{ this.Height = value; }-*/;
	
	// We cannot use instanceof to check the type of JavaScriptObject
	// the workaround is to compare with the instance field instead.

	public final native <T extends FormElement> boolean isSameTypeAs(FormElement instance)/*-{
		
		return (this.__type === instance.__type);
		
	}-*/;
}
